cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/UsageHandler.java

```diff
@@ -111,6 +120,111 @@ public void getDatasetAppUsage(HttpRequest request, HttpResponder responder,
                                  @PathParam("dataset-id") String datasetId) {
     final DatasetId id = new DatasetId(namespaceId, datasetId);
     Set<? extends EntityId> ids = registry.getPrograms(id);
-    responder.sendJson(HttpResponseStatus.OK, ids);
+    responder.sendJson(HttpResponseStatus.OK,
+                       Collections2.transform(ids, BackwardCompatibility.MAKE_COMPATIBLE_WITH_35));
+  }
+
+  /**
+   * Class to wrap stuff needed to make {@link UsageHandler} backward compatible. See CDAP-7316
+   * Also, see {@link IdCompatibleEntityIds} documentation for more details.
+   *
+   * Note: This backward compatibility transformation does not have its separate unit tests since since the unit test
+   * for {@link UsageHandler} will deserialize the JSON for the below classes to {@link Id} class and any
+   * incompatibility will fail the JSON deserialization there.
+   */
+  private static final class BackwardCompatibility {
+
+    /**
+     * Function which transforms and {@link EntityId} {@link IdCompatibleEntityIds} for backward compatibility with
+     * {@link Id}. This only supports {@link EntityId} which can be returned by {@link UsageHandler}.
+     */
+    private static final Function<EntityId, IdCompatibleEntityIds> MAKE_COMPATIBLE_WITH_35 =
+      new Function<EntityId, IdCompatibleEntityIds>() {
+        @Override
+        public IdCompatibleEntityIds apply(final EntityId entity) {
+          if (entity instanceof DatasetId) {
```

Comment:
This transform function is shared between all the Set<EntityId> if we tie it a specific one then we will need to write functions for each possible Set right ? We need to know what we are seeing so that we can create the right IdCompatibleEntityIds. Is there any other way to achieve this ?